### DISCUSSION TITLE
New Linux CosmosDb image much better but not ready when using Testcontainers

### DISCUSSION THREAD

**Discussion by diegosasw** [testcontainers/testcontainers-dotnet#1306] [Q&A]
Microsoft has released a new docker image which fixes many of the many issues they had.

https://devblogs.microsoft.com/cosmosdb/introducing-the-new-linux-based-azure-cosmos-db-emulator-preview/

I gave it a try manually and it works better, it loads way faster.
However I tried to use it as a test container by providing a `.WithImage(imageNameAndTag)`

And it doesn't work. It is as if test containers couldn't know when the container is up and running and it's stuck waiting.

I checked logs within container and its traces are very different than the previous versions' traces.

Does the current code need to change to support this image? Could it be changed to ensure it's backwards compatible?

**Comment by HofmeisterAn** [comment]
> It is as if test containers couldn't know when the container is up and running and it's stuck waiting.

If the current [wait strategy](https://github.com/testcontainers/testcontainers-dotnet/blob/37e5351ac206808601f6883ba7dd3999e5abcb46/src/Testcontainers.CosmosDb/CosmosDbBuilder.cs#L70-L96) is not compatible with the new (preview) tag, this is exactly what you’ll notice.

> Does the current code need to change to support this image? 

Yes, if the upcoming version contains breaking changes.

> Could it be changed to ensure it's backwards compatible?

I think the best we can do is figure out what changes are needed to be compatible with the upcoming update and are ready to publish a fix as soon as possible (at least provide guidance as a workaround, like overriding the default builder configuration).

**Comment by AButler** [comment]
This is how to get the container running:

```c#
_container = new CosmosDbBuilder()
                .WithImage("mcr.microsoft.com/cosmosdb/linux/azure-cosmos-emulator:vnext-preview")
                .WithCommand("--protocol", "https")
                .WithEnvironment("ENABLE_EXPLORER", "false")
                .WithWaitStrategy(Wait.ForUnixContainer().AddCustomWaitStrategy(new WaitUntil()))
                .Build();
```

The `WaitUtil` class is defined here (it is slightly different URL than the one in the Testcontainers repo):

```c#
private sealed class WaitUntil : IWaitUntil
{
    public async Task<bool> UntilAsync(IContainer container)
    {
        // CosmosDB's preconfigured HTTP client will redirect the request to the container.
        const string requestUri = "https://localhost";

        var httpClient = ((CosmosDbContainer)container).HttpClient;

        try
        {
            using var httpResponse = await httpClient.GetAsync(requestUri).ConfigureAwait(false);

            return httpResponse.IsSuccessStatusCode;
        }
        catch (Exception)
        {
            return false;
        }
        finally
        {
            httpClient.Dispose();
        }
    }
}
```

Note: this new emulator doesn't yet support everything the old one does.

**Reply by HofmeisterAn** [reply]
Cool, thanks. Did you check if [rewriting](https://github.com/testcontainers/testcontainers-dotnet/blob/49ef7dcb39e6212deb5809ae5e81eef59328c9dc/src/Testcontainers.CosmosDb/CosmosDbContainer.cs#L41-L65) the HTTP request is still necessary?

**Reply by AButler** [reply]
@HofmeisterAn Yes, the rewriting is still necessary.

**Reply by AButler** [reply]
To clarify the commands above,

```c#
.WithCommand("--protocol", "https")
```
is required to use the emulator with the .NET SDK (see https://learn.microsoft.com/en-gb/azure/cosmos-db/emulator-linux#running)

```c#
.WithEnvironment("ENABLE_EXPLORER", "false")
```
is optional as it doesn't start the UI for the Cosmos Explorer, so it speeds up startup time.

The different `WaitUntil` class is only different in that is checking `https://localhost` rather than `https://localhost/_explorer/emulator.pem`. This is because the Cosmos Explorer is no longer hosted at `/_explorer` and the certificate isn't exposed.

**Reply by NelsonBN** [reply]
I tested a different approach where I removed the override from DelegatingHandler. However, this requires knowing the random port beforehand to pass it as an environment variable during container setup. You can see an example [here](https://github.com/NelsonBN/demo-dotnet-integration-tests-cosmosdb-container/blob/main/tests/Api.PortOverrides.Tests/IntegrationTestsFactory.cs). After @HofmeisterAn shared this discussion, I now agree that rewriting the HTTP request is still the best approach.

I believe we don't need to create an costum IWaitUntil. However, we can use the following instead:

```csharp
.WithWaitStrategy(Wait.ForUnixContainer().UntilHttpRequestIsSucceeded(request => request.ForPort(CosmosDbPort)));
```

@AButler

The `.WithCommand("--protocol", "https")` is no longer necessary if we change the UriScheme to `Uri.UriSchemeHttp`.

```csharp
protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
{
    request.RequestUri = new UriBuilder(Uri.UriSchemeHttp, _hostname, _port, request.RequestUri.PathAndQuery).Uri;
    return base.SendAsync(request, cancellationToken);
}
```


And, the workaround for ServerCertificateCustomValidationCallback is no longer required.

```csharp
public UriRewriter(string hostname, ushort port)
    : base(new HttpClientHandler())
{
    _hostname = hostname;
    _port = port;
}
```

I also suggest delivering a properly configured CosmosClient because the emulator only supports connection gateway

```csharp
public CosmosClient CosmosClient
    => new CosmosClient(
        GetConnectionString(),
        new()
        {
            ConnectionMode = ConnectionMode.Gateway,
            HttpClientFactory = () => new(new UriRewriter(Hostname, GetMappedPublicPort(CosmosDbBuilder.CosmosDbPort)))
        });
```

[here an example](https://github.com/NelsonBN/demo-dotnet-integration-tests-cosmosdb-container/blob/main/tests/Api.DelegatingHandlerOverrides.Tests/IntegrationTestsFactory.cs)

**Reply by HofmeisterAn** [reply]
> I believe we don't need to create an costum IWaitUntil. However, we can use the following instead:

IIRC, this is flaky with the current version (indicate readiness too early), but OC could be different with the preview version.

> I also suggest delivering a properly configured CosmosClient because the emulator only supports connection gateway

Modules do not include client dependencies. We try to avoid dependency clashes.

**Reply by NelsonBN** [reply]
I ran some more intensive tests, and this new version (vnext-preview) still shows some flakiness. It seems much less frequent, but it still happens. However, the strategy suggested by @AButler doesn't fully solve the issue. I only achieved stability after adding a delay following a successful health check. With ~1000ms, I got 0% flakiness, running continuously for about 10minutes. I also opened an issue in the emulator's [repo](https://github.com/Azure/azure-cosmos-db-emulator-docker/issues/154) to ask if there's a reliable way to ensure when the API is fully ready.

Here's the implementation that gave me the best results

```csharp
internal class WaitUntil : IWaitUntil
{
    public async Task<bool> UntilAsync(IContainer container)
    {
        // CosmosDB's preconfigured HTTP client will redirect the request to the container.
        const string REQUEST_URI = "http://localhost";

        using var client = new HttpClient(new UriRewriter(container.Hostname, container.GetMappedPublicPort(8081)));


        try
        {
            using var httpResponse = await client.GetAsync(REQUEST_URI).ConfigureAwait(false);

            if(httpResponse.IsSuccessStatusCode)
            {
                await Task.Delay(1_000);
                return true;
            }
        }
        catch { }
        return false;
    }
}
```

[Here's](https://github.com/NelsonBN/demo-dotnet-integration-tests-cosmosdb-container/blob/main/tests/Readiness.Tests/Program.cs) my tests


@HofmeisterAn Regarding the CosmosDB client dependency, the idea was to provide a properly configured client, since using the emulator requires overriding the `HttpClientFactory` and setting `ConnectionMode = ConnectionMode.Gateway`. I understand the concern about avoiding external dependencies, but since .NET always uses the Azure CosmosDB client, I think it could be a good idea to deliver a client that's already correctly configured.

**Reply by HofmeisterAn** [reply]
> However, the strategy suggested by @AButler doesn't fully solve the issue. I only achieved stability after adding a delay following a successful health check. With ~1000ms, I got 0% flakiness, running continuously for about 10minutes.

This sounds unfortunate. I hope we get some guidance from upstream. The ~1000ms may still be flaky depending on the hardware and environment, which is not ideal.

> @HofmeisterAn Regarding the CosmosDB client dependency, the idea was to provide a properly configured client, since using the emulator requires overriding the `HttpClientFactory` and setting `ConnectionMode = ConnectionMode.Gateway`. I understand the concern about avoiding external dependencies, but since .NET always uses the Azure CosmosDB client, I think it could be a good idea to deliver a client that's already correctly configured.

This applies to almost every module. As mentioned earlier, we are not providing something like that on purpose. I suggest adding module-specific documentation instead.

**Reply by AButler** [reply]
> The .WithCommand("--protocol", "https") is no longer necessary if we change the UriScheme to Uri.UriSchemeHttp.

The reason this was added was as per the documentation at: https://learn.microsoft.com/en-gb/azure/cosmos-db/emulator-linux#running

Screenshot here: 
![image](https://github.com/user-attachments/assets/c4fda274-51b6-46e0-85bb-ef1f5b23af32)

**Reply by jundayin** [reply]
Have anyone tried out the latest vnext release and see if the integration improved?
